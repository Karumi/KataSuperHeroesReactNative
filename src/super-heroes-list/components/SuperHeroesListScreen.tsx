import * as React from "react";
import { Text } from "react-native";
import navigationOptions from "../../base-components/navigationOptions";
import { SuperHero } from "../../core/model";
import SuperHeroCell from "./SuperHeroCell";

const sh = new SuperHero("Scarlet Witch",
    "https://i.annihil.us/u/prod/marvel/i/mg/9/b0/537bc2375dfb9.jpg",
    true,
    "Scarlet Witch was born at the Wundagore base of the High Evolutionary, she and her twin "
    + "brother Pietro were the children of Romani couple Django and Marya Maximoff. The "
    + "High Evolutionary supposedly abducted the twins when they were babies and "
    + "experimented on them, once he was disgusted with the results, he returned them to"
    + " Wundagore, disguised as regular mutants.");

class SuperHeroesListScreen extends React.Component {
    public static navigationOptions = navigationOptions("Kata Super Heroes");
    public render() {
        return <SuperHeroCell superHero={sh} />;
    }
}

export default SuperHeroesListScreen;
