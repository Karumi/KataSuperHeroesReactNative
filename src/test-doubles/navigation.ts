
export const mockPlainNavigation = {
    navigate: jest.fn(),
};

export const navigationWithMockedParams = {
    navigate: jest.fn(),
    getParam: jest.fn(),
};
