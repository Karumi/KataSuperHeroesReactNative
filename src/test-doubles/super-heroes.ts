import { SuperHero } from "../core/model";

export const anySuperHeroId: string = "Gambito";

export const anyListOfSuperHeroes: SuperHero[] = [
    new SuperHero("Scarlet Witch",
        "https://i.annihil.us/u/prod/marvel/i/mg/9/b0/537bc2375dfb9.jpg",
        false,
        "Scarlet Witch was born at the Wundagore base of the High Evolutionary, she and her twin "
        + "brother Pietro were the children of Romani couple Django and Marya Maximoff. The "
        + "High Evolutionary supposedly abducted the twins when they were babies and "
        + "experimented on them, once he was disgusted with the results, he returned them to"
        + " Wundagore, disguised as regular mutants."),
    new SuperHero("Iron Man",
        "https://i.annihil.us/u/prod/marvel/i/mg/c/60/55b6a28ef24fa.jpg",
        true,
        "Wounded, captured and forced to build a weapon by his enemies, billionaire "
        + "industrialist Tony Stark instead created an advanced suit of armor to save his "
        + "life and escape captivity. Now with a new outlook on life, Tony uses his money "
        + "and intelligence to make the world a safer, better place as Iron Man."),
    new SuperHero("Wolverine",
        "https://i.annihil.us/u/prod/marvel/i/mg/9/00/537bcb1133fd7.jpg",
        false,
        "Born with super-human senses and the power to heal from almost any wound, " +
        "Wolverine "
        + "was captured by a secret Canadian organization and given an unbreakable "
        + "skeleton and claws. Treated like an animal, it took years for him to control"
        + " himself. Now, he's a premiere member of both the X-Men and the Avengers."),
];

export const anyRegularSuperHero = anyListOfSuperHeroes[0];

export const anyAvenger = anyListOfSuperHeroes[1];
