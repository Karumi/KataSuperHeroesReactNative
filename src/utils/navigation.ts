export interface NavigationLeafRoute<Params> {
    params?: Params;
  }
