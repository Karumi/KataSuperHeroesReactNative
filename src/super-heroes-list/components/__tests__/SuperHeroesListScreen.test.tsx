import * as React from "react";
import { Provider } from "react-redux";
import * as renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { action } from "typesafe-actions";
import { mockPlainNavigation } from "../../../test-doubles/navigation";
import { anyListOfSuperHeroes } from "../../../test-doubles/super-heroes";
import SuperHeroesListScreen from "../SuperHeroesListScreen";

jest.mock("../../../base-components/loading/Loading", () => "Loading");

jest.mock("../../../base-components/empty-case/EmptyCase", () => "EmptyCase");

jest.mock("../SuperHeroesList", () => "SuperHeroesList");

const anyFetchSuperHeroesAction = action("fetchSuperHeroes");

jest.mock("../../super-heroes-list-actions", () => {
    return {
        fetchSuperHeroes: (() => anyFetchSuperHeroesAction),
    };
});

it("renders a loading component if the users list state is loading", () => {
    const store = givenAStoreInTheLoadingState();

    const component = renderer
        .create(
            <Provider store={store} >
                <SuperHeroesListScreen navigation={mockPlainNavigation} />
            </Provider>,
        )
        .toJSON();
    expect(component).toMatchSnapshot();
});

it("renders an empty case component if the list of super heroes is empty", () => {
    const store = givenAStoreLoadedButWithoutSuperHeroes();

    const component = renderer
        .create(
            <Provider store={store} >
                <SuperHeroesListScreen navigation={mockPlainNavigation} />
            </Provider>,
        )
        .toJSON();
    expect(component).toMatchSnapshot();
});

it("renders the list of super heroes if we are not fetching the info and we have some super heroes to show", () => {
    const store = givenAStoreLoadedWithSuperHeroes();

    const component = renderer
        .create(
            <Provider store={store} >
                <SuperHeroesListScreen navigation={mockPlainNavigation} />
            </Provider>,
        )
        .toJSON();
    expect(component).toMatchSnapshot();
});

it("invokes the onMount prop when the component is mounted", () => {
    const store = givenAStoreLoadedWithSuperHeroes();

    renderer.create(
        <Provider store={store}>
            <SuperHeroesListScreen navigation={mockPlainNavigation} />
        </Provider>,
    );

    const actions = store.getActions();
    expect(actions).toEqual([anyFetchSuperHeroesAction]);
});

function givenAStoreInTheLoadingState() {
    const middlewares = [thunk];
    return configureStore(middlewares)({
        superHeroesList: {
            loading: true,
            superHeroes: [],
        },
    });
}

function givenAStoreLoadedButWithoutSuperHeroes() {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);
    return mockStore({
        superHeroesList: {
            loading: false,
            superHeroes: [],
        },
    });
}

function givenAStoreLoadedWithSuperHeroes() {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);
    return mockStore({
        superHeroesList: {
            loading: false,
            superHeroes: anyListOfSuperHeroes,
        },
    });
}
