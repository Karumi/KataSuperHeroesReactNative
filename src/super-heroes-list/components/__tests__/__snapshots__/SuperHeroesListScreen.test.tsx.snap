// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a loading component if the users list state is loading 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#22282F",
      "flex": 1,
    }
  }
>
  <Loading />
</View>
`;

exports[`renders an empty case component if the list of super heroes is empty 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#22282F",
      "flex": 1,
    }
  }
>
  <EmptyCase />
</View>
`;

exports[`renders the list of super heroes if we are not fetching the info and we have some super heroes to show 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#22282F",
      "flex": 1,
    }
  }
>
  <SuperHeroesList
    onSuperHeroCellTap={[Function]}
    superHeroes={
      Array [
        SuperHero {
          "description": "Scarlet Witch was born at the Wundagore base of the High Evolutionary, she and her twin brother Pietro were the children of Romani couple Django and Marya Maximoff. The High Evolutionary supposedly abducted the twins when they were babies and experimented on them, once he was disgusted with the results, he returned them to Wundagore, disguised as regular mutants.",
          "isAvenger": false,
          "name": "Scarlet Witch",
          "picture": "https://i.annihil.us/u/prod/marvel/i/mg/9/b0/537bc2375dfb9.jpg",
        },
        SuperHero {
          "description": "Wounded, captured and forced to build a weapon by his enemies, billionaire industrialist Tony Stark instead created an advanced suit of armor to save his life and escape captivity. Now with a new outlook on life, Tony uses his money and intelligence to make the world a safer, better place as Iron Man.",
          "isAvenger": true,
          "name": "Iron Man",
          "picture": "https://i.annihil.us/u/prod/marvel/i/mg/c/60/55b6a28ef24fa.jpg",
        },
        SuperHero {
          "description": "Born with super-human senses and the power to heal from almost any wound, Wolverine was captured by a secret Canadian organization and given an unbreakable skeleton and claws. Treated like an animal, it took years for him to control himself. Now, he's a premiere member of both the X-Men and the Avengers.",
          "isAvenger": false,
          "name": "Wolverine",
          "picture": "https://i.annihil.us/u/prod/marvel/i/mg/9/00/537bcb1133fd7.jpg",
        },
      ]
    }
  />
</View>
`;
